java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql

```diff
@@ -0,0 +1,164 @@
+  MDUpdateMethod() {
+    this.getDeclaringType() instanceof MessageDigest and
+    this.hasName("update")
+  }
+}
+
+/** The hashing method that could taint the input. */
+class MDHashMethodAccess extends MethodAccess {
+  MDHashMethodAccess() {
+    (
+      this.getMethod() instanceof MDDigestMethod or
+      this.getMethod() instanceof MDUpdateMethod
+    ) and
+    this.getNumArgument() != 0
+  }
+}
+
+/** Gets a regular expression for matching common names of variables that indicate the value being held is a password. */
+string getPasswordRegex() { result = "(?i).*pass(wd|word|code|phrase).*" }
+
+/** Finds variables that hold password information judging by their names. */
+class PasswordVarExpr extends Expr {
+  PasswordVarExpr() {
+    this.(VarAccess).getVariable().getName().toLowerCase().regexpMatch(getPasswordRegex()) and
+    not this.(VarAccess).getVariable().getName().toLowerCase().matches("%hash%") // Exclude variable names such as `passwordHash` since their values were already hashed
+  }
+}
+
+/** Holds if `Expr` e is an operand of `AddExpr`. */
+predicate hasAddExpr(AddExpr ae, Expr e) {
+  ae.getAnOperand() = e or
+  hasAddExpr(ae.getAnOperand(), e)
+}
+
+/** Holds if `MethodAccess` ma has a flow to another `MDHashMethodAccess` call. */
+predicate hasAnotherHashCall(MethodAccess ma) {
+  exists(MethodAccess ma2, DataFlow2::Node node1, DataFlow2::Node node2 |
+    ma2 instanceof MDHashMethodAccess and
+    ma2 != ma and
+    node1.asExpr() = ma.getAChildExpr() and
+    node2.asExpr() = ma2.getAChildExpr() and
+    (
+      TaintTracking2::localTaint(node1, node2) or
+      TaintTracking2::localTaint(node2, node1)
+    )
+  )
+}
+
+/** Holds if `MethodAccess` ma is a hashing call without a sibling node making another hashing call. */
+predicate isSingleHashMethodCall(MethodAccess ma) {
+  (
+    ma instanceof MDHashMethodAccess and
+    not hasAnotherHashCall(ma)
+  )
+}
+
+/** Holds if `MethodAccess` ma is invoked by `MethodAccess` ma2 either directly or indirectly. */
+predicate hasParentCall(MethodAccess ma2, MethodAccess ma) {
+  ma.getCaller() = ma2.getMethod() and
+  not ma2 instanceof MDHashMethodAccess
+  or
+  exists(MethodAccess ma3 |
+    ma.getCaller() = ma3.getMethod() and
+    not ma3 instanceof MDHashMethodAccess and
+    hasParentCall(ma2, ma3)
+  )
+}
+
+/** Holds if `MethodAccess` is a single hashing call. */
+predicate isSink(MethodAccess ma) {
+  isSingleHashMethodCall(ma) and
+  not exists(MethodAccess ma2 | hasParentCall(ma2, ma))
+}
+
+/** Sink of hashing calls. */
+class HashWithoutSaltSink extends DataFlow::ExprNode {
+  HashWithoutSaltSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getAnArgument() and
+      isSink(ma)
+    )
+  }
+}
+
+/** Taint configuration tracking flow from an expression whose name suggests it holds password data to a method call that generates a hash without a salt. */
+class HashWithoutSaltConfiguration extends TaintTracking::Configuration {
```

Comment:
Since we now only have one Configuration I guess all DataFlow2,3,... can be discarded from this PR

Reply:
Removed <code>DataFlow2</code> and the newly added <code>TaintTracking3</code> since they are no longer being used.